### PR TITLE
Centauri monorepo

### PR DESCRIPTION
**Pull request by seunlanlege**
This consolidates all our light clients, light client verification functions, ibc implementations and relayer under the same repo for ease of integration testing.

### CODE REVIEW

File: light-clients/ics07-tendermint/src/client_state.rs [hunk 1]
```diff
@@ -0,0 +1,677 @@
...
+				.trust_level
+				.try_into()
+				.map_err(|e: Ics02Error| Error::invalid_trust_threshold(e.to_string()))?,
+			trusting_period: self.trusting_period,
+			clock_drift: self.max_clock_drift,
+		})
+	}
+
+	/// Verify the time and height delays
+	pub fn verify_delay_passed(
```
Comment (Wizdave97) [inline light-clients/ics07-tendermint/src/client_state.rs:195]: we can move this function to the `common` lib, it's implementation would be the same for all the light clients.

File: light-clients/ics10-grandpa/src/client_state.rs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	pub public_key: Public,
+	/// authority weight
+	pub weight: u64,
+}
+
+#[derive(PartialEq, Clone, Debug, Default, Eq)]
+pub struct ClientState<H> {
+	/// Relay chain
+	pub relay_chain: RelayChain,
+	/// Latest mmr root hash
```
Comment (Wizdave97) [inline light-clients/ics10-grandpa/src/client_state.rs:36]:
```suggestion
	/// Latest relay chain block hash
```

File: light-clients/ics10-grandpa/src/proto/grandpa.proto [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+message Authority {
+  // ed25519 public key of the authority
+  bytes public_key = 1;
+  // authority weight
+  uint64 weight = 2;
+}
+
+// ClientState for the grandpa client
+message ClientState {
+  // Latest mmr root hash
```
Comment (Wizdave97) [inline light-clients/ics10-grandpa/src/proto/grandpa.proto:22]:
```suggestion
  // Latest relay chain block hash
```

File: light-clients/ics10-grandpa/src/proto/grandpa.proto [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  bytes latest_relay_hash = 1;
+
+  // block number for the latest mmr_root_hash
```
Comment (Wizdave97) [inline light-clients/ics10-grandpa/src/proto/grandpa.proto:25]:
```suggestion
  // current grandpa authority set id
```

File: light-clients/ics10-grandpa/src/proto/grandpa.proto [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  // Known relay chains
+  RelayChain relay_chain = 4;
+
+  // ParaId of associated parachain
+  uint32 para_id = 5;
+
+  // latest parachain height
+  uint32 latest_para_height = 6;
+
+  // latest parachain height
```
Comment (Wizdave97) [inline light-clients/ics10-grandpa/src/proto/grandpa.proto:40]:
```suggestion
  // Current grandpa authorities
```

**Review by Wizdave97 — COMMENTED** [review]
Great work on this

File: algorithms/beefy/src/lib.rs [hunk 6]
```diff
@@ -204,10 +205,15 @@ where
...
 
 		// Verify timestamp extrinsic
-		let proof = &*parachain_header.extrinsic_proof;
-		let ext = &*parachain_header.timestamp_extrinsic;
-		let extrinsic_root = decoded_para_header.extrinsics_root;
-		<H as HostFunctions>::verify_timestamp_extrinsic(extrinsic_root, proof, ext)?;
+		// Timestamp extrinsic should be the first inherent and hence the first extrinsic
+		// https://github.com/paritytech/substrate/blob/d602397a0bbb24b5d627795b797259a44a5e29e9/primitives/trie/src/lib.rs#L99-L101
+		let timestamp_ext_key = codec::Compact(0u32).encode();
+		sp_trie::verify_trie_proof::<LayoutV0<H::BlakeTwo256>, _, _, _>(
```
Comment (Wizdave97) [inline algorithms/beefy/src/lib.rs:211]: Shouldn't we use the host function, `sp_io::trie::blake2_256_verify_proof` here ?

**Review by Wizdave97 — CHANGES_REQUESTED** [review]

File: algorithms/grandpa/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+			.map_err(|err| anyhow!("error verifying parachain header state proof: {err}"))?
+			.remove(key.as_ref())
+			.flatten()
+			.ok_or_else(|| anyhow!("Invalid proof, parachain header not found"))?;
+			let header = Vec::<u8>::decode(&mut &header[..])?;
+			let parachain_header = H::decode(&mut &header[..])?;
+			// Timestamp extrinsic should be the first inherent and hence the first extrinsic
+			// https://github.com/paritytech/substrate/blob/d602397a0bbb24b5d627795b797259a44a5e29e9/primitives/trie/src/lib.rs#L99-L101
+			let key = codec::Compact(0u32).encode();
+			sp_trie::verify_trie_proof::<LayoutV0<Host::BlakeTwo256>, _, _, _>(
```
Comment (Wizdave97) [inline algorithms/grandpa/src/lib.rs:89]: Also here
```suggestion
			sp_trie::verify_trie_proof::<LayoutV0<Host::BlakeTwo256>, _, _, _>(
```

File: algorithms/beefy/src/lib.rs (hunk 6 above)
Comment (seunlanlege) [reply algorithms/beefy/src/lib.rs:211]: Yeah that wouldn't be chain agnostic, remember we have to support near and cosmos (via ics-8 wasm client)

Comment (Wizdave97) [reply algorithms/beefy/src/lib.rs:211]: Okay, makes sense.

Comment (seunlanlege) [reply algorithms/beefy/src/lib.rs:211]: The `H::BlakeTwo256: hash_db::Hashing` so we can implement it using the different host functions for the different runtimes

**Review by Wizdave97 — APPROVED** [review]

**Pull request merged**